fdbclient/ManagementAPI.actor.cpp

```diff
@@ -2115,34 +2117,79 @@ ACTOR Future<Void> lockDatabase(Transaction* tr, UID id) {
 		}
 	}
 
-	tr->atomicOp(databaseLockedKey,
+	tr->atomicOp(lockKey,
 	             BinaryWriter::toValue(id, Unversioned()).withPrefix("0123456789"_sr).withSuffix("\x00\x00\x00\x00"_sr),
 	             MutationRef::SetVersionstampedValue);
-	tr->addWriteConflictRange(normalKeys);
+	tr->addWriteConflictRange(affectedRange);
 	return Void();
 }
 
-ACTOR Future<Void> lockDatabase(Reference<ReadYourWritesTransaction> tr, UID id) {
+ACTOR template <class Tr>
+Future<Void> unlockDBOrTenant(Tr* tr, UID id, StringRef lockKey) {
 	tr->setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
 	tr->setOption(FDBTransactionOptions::LOCK_AWARE);
-	Optional<Value> val = wait(tr->get(databaseLockedKey));
+	Optional<Value> val = wait(tr->get(lockKey));
 
-	if (val.present()) {
-		if (BinaryReader::fromStringRef<UID>(val.get().substr(10), Unversioned()) == id) {
-			return Void();
-		} else {
-			//TraceEvent("DBA_LockLocked").detail("Expecting", id).detail("Lock", BinaryReader::fromStringRef<UID>(val.get().substr(10), Unversioned()));
-			throw database_locked();
-		}
+	if (!val.present())
+		return Void();
+
+	if (val.present() && BinaryReader::fromStringRef<UID>(val.get().substr(10), Unversioned()) != id) {
+		//TraceEvent("DBA_UnlockLocked").detail("Expecting", id).detail("Lock", BinaryReader::fromStringRef<UID>(val.get().substr(10), Unversioned()));
+		throw database_locked();
```

Comment:
Is there going to be a different error for the tenant case at some point (e.g. `tenant_locked`)?

Reply:
I want to do code sharing -- locking a db or a tenant uses the same code, so I'd like to only have one actor for this as opposed to two. Alternatively I can move it to a different set of files (like `Locking.actor.h` and `Locking.actor.cpp`)

Reply:
I think it will be valuable to have a different error here, since the implications of both are different and we expect clients to actually encounter and handle a `tenant_locked` type error. This should still be possible with shared code, though.

Reply:
As far as my suggestion to move the tenant functions out, I think you can leave this function here and only move `lockTenant`, etc.